aspnetcore/mvc/controllers/application-model.md

```diff
@@ -12,7 +12,184 @@ ms.technology: aspnet
+
+[!code-csharp[Main](./application-model/sample/src/AppModelSample/Conventions/CustomActionNameAttribute.cs)]
+
+This attribute is applied to an action method in the `HomeController`:
+
+[!code-csharp[Main](./application-model/sample/src/AppModelSample/Controllers/HomeController.cs?name=ActionModelConvention&highlight=2)]
+
+Even though the method name is `SomeName`, the attribute overrides the MVC convention of using the method name and replaces the action name with `MyCoolAction`. Thus, the route used to reach this action is `/Home/MyCoolAction`.
+
+### Sample: Custom Routing Convention
+
+You can use an `IApplicationModelConvention` to customize how routing works. For example, the following convention will incorporate Controllers' namespaces into their routes, replacing `.` in the namespace with `/` in the route:
+
+[!code-csharp[Main](./application-model/sample/src/AppModelSample/Conventions/NamespaceRoutingConvention.cs)]
+
+The convention is added as an option in Startup.
+
+[!code-csharp[Main](./application-model/sample/src/AppModelSample/Startup.cs?name=ConfigureServices&highlight=6)]
+
+> [!TIP]
+> If you need to add a convention to MVC separately from where MVC is configured (for instance, as part of your own [middleware](https://docs.microsoft.com/aspnet/core/fundamentals/middleware)), you can access `MvcOptions` using `services.Configure<MvcOptions>(c => c.Conventions.Add(YOURCONVENTION));`
+
+This sample only applies to convention to routes that aren't using attribute routing, and for demo purposes it's further constrained to only apply to controllers with "Namespace" in the name. This controller demonstrates the behavior applied by this convention:
+
+[!code-csharp[Main](./application-model/sample/src/AppModelSample/Controllers/NamespaceRoutingController.cs?highlight=7-8)]
+
+## Application Model Usage in WebApiCompatShim
+
+ASP.NET MVC Core uses a different set of conventions from ASP.NET Web API 2. Using custom conventions, you can modify an ASP.NET MVC Core app's behavior to be consistent with that of a Web API app. Microsoft ships the [WebApiCompatShim](https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc.WebApiCompatShim/) specifically for this purpose.
+
+> [!NOTE]
+> Learn more about [migrating from ASP.NET Web API](https://docs.microsoft.com/aspnet/core/migration/webapi)).
+
+To use the Web API Compatibility Shim, you need to add the package to your project and then add the conventions to MVC by calling `AddWebApiConventions` in `Startup`:
+
+```c#
+services.AddMvc().AddWebApiConventions();
+
+```
+
+The conventions provided by the shim are only applied to parts of the app that have had certain attributes applied to them. The following four attributes are used to control which controllers should have their conventions modified by the shim's conventions:
+
+* [UseWebApiActionConventionsAttribute]()
```

Comment:
Those were all supposed to be links - I just forgot to go back and add them. Fixed.